### PR TITLE
Matt Burns - Assignment 3

### PR DESCRIPTION
**Pull request by m6urns**
## Data Loader Implementation
The data loader implementation consists of two main components: a custom StarWarsDataset class and a create_dataloaders utility function. This approach was chosen for its flexibility and clean separation of concerns:

The StarWarsDataset class extends PyTorch's Dataset class to handle Star Wars dialogue data specifically. It uses the BERT tokenizer for text processing, which is ideal for NLP tasks. The implementation includes automatic padding and truncation to ensure consistent sequence lengths, making it suitable for batch processing. The dataset returns dictionaries containing input_ids, attention_masks, and character labels, providing all necessary components for training a model.

The create_dataloaders function handles the creation of both training and validation data loaders with configurable batch sizes and split ratios.

## Testing Strategy
The test suite was designed to verify both the functionality and robustness of the data loader implementation. Key test categories include:

1. Structural Tests:
   - test_dataset_structure: Verifies the base dataset structure and required fields
   - test_dataset_initialization: Ensures proper dataset initialization
   - test_dataset_item_format: Validates the format of individual items

2. Constraint Tests:
   - test_max_length_constraint: Verifies sequence length constraints
   - test_attention_mask_validity: Ensures attention masks contain only valid values (0s and 1s)

3. DataLoader Tests:
   - test_dataloader_creation: Verifies proper DataLoader configuration
   - test_batch_format: Ensures correct batch formatting
   - test_train_val_split: Validates the train/validation split ratio
   - test_various_batch_sizes: Uses parametrized testing to verify functionality across different batch sizes

## Potential Improvements
Given more time, I would make the following improvements:

1. Performance Optimization:
   - Implement multi-worker data loading for better performance on larger datasets
   - Add caching mechanisms for tokenized outputs to reduce preprocessing overhead

2. Enhanced Testing:
   - Add edge case testing for very long sequences
   - Add integration tests with the full ZenML pipeline

### CODE REVIEW

File: Assignment3.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+2. Constraint Tests:
+   - test_max_length_constraint: Verifies sequence length constraints
+   - test_attention_mask_validity: Ensures attention masks contain only valid values (0s and 1s)
+
+3. DataLoader Tests:
+   - test_dataloader_creation: Verifies proper DataLoader configuration
+   - test_batch_format: Ensures correct batch formatting
+   - test_train_val_split: Validates the train/validation split ratio
+   - test_various_batch_sizes: Uses parametrized testing to verify functionality across different batch sizes
```
Comment (msharp9) [inline Assignment3.md:26]: I'm happy for you to use an LLM to help write your response, but all of this is already clear from the code/docstrings.  It doesn't answer the question of "why you wrote the tests you did" it just tells me what you did.

File: Assignment3.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+## Potential Improvements
+Given more time, I would make the following improvements:
+
+1. Performance Optimization:
+   - Implement multi-worker data loading for better performance on larger datasets
+   - Add caching mechanisms for tokenized outputs to reduce preprocessing overhead
+
+2. Enhanced Testing:
+   - Add edge case testing for very long sequences
+   - Add integration tests with the full ZenML pipeline
```
Comment (msharp9) [inline Assignment3.md:37]: This is beyond the scope of this class.  Simply running it, and taking a screenshot to prove the pipeline runs is good enough.

File: Assignment3.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+   - test_dataloader_creation: Verifies proper DataLoader configuration
+   - test_batch_format: Ensures correct batch formatting
+   - test_train_val_split: Validates the train/validation split ratio
+   - test_various_batch_sizes: Uses parametrized testing to verify functionality across different batch sizes
+
+## Potential Improvements
+Given more time, I would make the following improvements:
+
+1. Performance Optimization:
+   - Implement multi-worker data loading for better performance on larger datasets
```
Comment (msharp9) [inline Assignment3.md:32]: It's your job to make sure suggestions from an LLM make sense before putting them down as your own.  As a note, multiprocessing comes for free with a pytorch dataloader, you just have to pass a num_workers arg.

File: tests/test_dataloader.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
+import pytest
+import torch
+from torch.utils.data import DataLoader
+from datasets import load_dataset
+import sys
+from pathlib import Path
+
+sys.path.append(str(Path(__file__).parent.parent))
```
Comment (msharp9) [inline tests/test_dataloader.py:8]: Is this needed?

File: tests/test_dataloader.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+from pipeline import StarWarsDataset, create_dataloaders
+
+
+def test_dataset_structure():
+    """Test that we can access the dataset and its expected structure."""
+    dataset = load_dataset("andrewkroening/Star-wars-scripts-dialogue-IV-VI")
```
Comment (msharp9) [inline tests/test_dataloader.py:15]: We should avoiding hitting any APIs in a test.  If Huggingface is down, this test will fail in confusing ways.  It's better to just create an test dataset.  Your only testing the very first item in the entire dataset anyways, no reason to pull the whole thing, which also makes it unnecessarily slow.

File: tests/test_dataloader.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    assert "train" in dataset
+    sample = dataset["train"][0]
+    assert "Line" in sample
+    assert "Character" in sample
+
+
+def test_dataset_initialization():
+    """Test that the dataset can be initialized properly."""
+    dataset = StarWarsDataset()
```
Comment (msharp9) [inline tests/test_dataloader.py:24]: Beyond the scope of this class, but how you have this class written and given my previous comment, you'd want to mock the load_dataset function before calling this to avoid hitting the Huggingface API.

**Review by msharp9 — COMMENTED** [review]
Very nice! Your code is clean and I just love how many tests you created to test everything.

Some notes on the explain your work section. A shorter but more meaningful response would have been better.

Also, it looks like you didn't `git pull` before creating your branch. Unfortunately, because of this your branch is behind main, and thus didn't contain the CI/CD pipeline, ensuring the pipeline passed was part of the assignment.
